### PR TITLE
Add a user command to close registereditor windows

### PR DESCRIPTION
**Pull request by nullromo**
The command is `:RegisterEditClose`. With no arguments, it closes all the registereditor buffers. Optionally, you can specify one or more registers and all corresponding registereditor buffers will be closed.

For instance, `:RegisterEdit a a a b c a` followed by `:RegisterEditClose a` would leave you with only 2 registereditor windows open (one for `b` and one for `c`).

### CODE REVIEW

**Comment by tuurep** [comment]
Hi sorry it took me a little long to look at this,

Interesting but I have some objections to this:

1. I've always used the command `:RegisterEdit` as `:R`, which works when there are no other commands sharing the same prefix, so if `:RegisterEditClose` exists, it gives this error:
   ```
   E464: Ambiguous use of user-defined command
   ```

2. I feel this mostly addresses a case where we as plugin devs open a lot of registereditor splits to test edge cases and such, but may not be very practical in real use of the plugin. Would you disagree?

3. It's quite a bit of code complexity for something I had used this builtin alternative myself:
   - Switch to main buffer and do [`<C-w>o`](https://neovim.io/doc/user/windows.html#CTRL-W_o) (or `:only`)
   - That doesn't allow closing only certain reg-splits though, but I have a feeling this could be done fairly easily with some light scripting in personal configs? (I'll look into that a bit more)

---

All that said I gave this branch a try and for some reason the command `:RegisterEditClose` isn't working for me:

https://github.com/user-attachments/assets/c971828a-0549-4653-afc2-ff8d06de08ac

**Comment by nullromo** [comment]
### Responses

> no other commands sharing the same prefix

It's not really realistic to assume this will be true. It sounds like maybe you don't use very many plugins or don't have many user commands set up. For example, when I press `:r<TAB>` I get a list showing

```
RegisterEdit
Registers
Rename
RenameFile
ResetCashRegisters
Ruler
```

These are all commands that I have set up.

> may not be very practical in real use

Yes, that's probably true.

> Switch to main buffer and do `<C-w>o`

Main buffer? Wouldn't this close all the other windows in the tab?

> isn't working for me

I messed up the rebase a little. Fixed now.

### Conclusion

I think this command is useful at the minimum as an API option. What I mean is that people should be able to call it from lua with something like `require('registereditor').close_buffers()` or something. It should be exposed like that. But as a user command, I agree that we shouldn't pollute the command space.

The best option may be to have it be an argument based command. Just check if the first token is an exact string and use that as the "action." `open` is the default action.

- `:RegisterEdit a` = edit register `a`
- `:RegisterEdit a b` = edit registers `a` and `b`
- `:RegisterEdit close a` = close buffers for register `a`
- `:RegisterEdit close a b` = close buffers for registers `a` and `b`
- `:RegisterEdit close` = close all registereditor buffers
- `:RegisterEdit` = error (argument required)
- `:RegisterEdit asdf` = error (asdf is not a valid command)
- `:RegisterEdit open a` = same as `:RegisterEdit a`

How does that sound?

**Comment by nullromo** [comment]
To test this version:
1. `:RegisterEdit a b c` should open 3 windows.
2. `:RegisterEdit open a b c` should open 3 more windows.
3. `:RegisterEdit close a` should close 2 of the windows.
4. `:RegisterEdit close` should close the remaining 4 windows.

**Comment by tuurep** [comment]
Yeah this is looking quite neat

I was investigating whether it's possible to do something like `:bd <prefix>` with builtins, but:
- doesn't seem so simple
- let's say we allow the buffer name to be something like `Editing register <reg>`, this solution now makes way more sense



> > no other commands sharing the same prefix
>
> It's not really realistic to assume this will be true

Yes you're right about this, I can only use `:R` as sort of a lucky coincidence.

> Switch to main buffer and do `<C-w>o`
>
> Main buffer? Wouldn't this close all the other windows in the tab?

Good point, `<C-w>o` only works for me because I almost never use more than one window split.

## Main command name

I'm feeling like we could shorten the main command name while also improving the semantics of the main + subcommands, I would suggest:

Main command `:Reg`, subcommands `edit` (or empty) and `close`:
- `:Reg a b c`
- `:Reg edit a b c`
- `:Reg close a`

What do you think, could this possibly conflict with another plugin?

I'm a little curious what you have on `:Registers` btw :smile: 

I find myself wanting to tab-complete the subcommand `close` but inserting a `^I` by accident instead. The [DOs and DONT's guide](https://github.com/nvim-neorocks/nvim-best-practices?tab=readme-ov-file#white_check_mark-do-1) had something about this, wonder if it's worth the effort?

## Testing

On testing, this is working good, only I notice that the error message is noisier than it used to be for non-valid arguments, like this:

image

**Comment by tuurep** [comment]
I can update the README after this PR btw, no worries about that

**Comment by tuurep** [comment]
Hi taken me too long but I'm gonna look at these two PRs now, code and all.

One question, is this suggestion on the command names ok?

> Final command names I think should be:
>
> - `:RegisterEditor open` (or empty)
> - `:RegisterEditor close`

File: lua/internals.lua [hunk 1]
```diff
@@ -104,26 +110,108 @@ local function check_string_is_register(value)
...
+                    should_close = true
+                end
+            end
+
+            -- close the buffer if necessary
+            if should_close then
+                close_buffer(buffer)
+            end
+        end
+    end)
```
Comment (tuurep) [inline lua/internals.lua:199]: We can possibly simplify this quite a bit using `vim.tbl_contains` like this:

```lua
    -- loop over all the buffers and close the appropriate ones.
    loop_over_register_buffers(function(buffer)
        local buffer_register = get_register_from_buffer(buffer)

        -- if the registers list is nil or empty, then close all the buffers
        local no_arguments = registers == nil or #registers == 0
        local should_close = no_arguments or vim.tbl_contains(registers, buffer_register)

        if should_close then
            close_buffer(buffer)
        end
    end)
```

**Review by tuurep — CHANGES_REQUESTED** [review]

**Comment by tuurep** [comment]
Cool, thanks! More to come later.

**Comment by tuurep** [comment]
Happy travels, no hurry 😄

File: plugin/registereditor.lua [hunk 2]
```diff
@@ -2,7 +2,7 @@ local internals = require("internals")
 
 local function setup_user_commands()
     vim.api.nvim_create_user_command("RegisterEdit", function(opts)
```
Comment (tuurep) [inline plugin/registereditor.lua:4]:
```suggestion
    vim.api.nvim_create_user_command("RegisterEditor", function(opts)
```

File: plugin/registereditor.lua [hunk 3]
```diff
@@ -2,7 +2,7 @@ local internals = require("internals")
...
-        internals.open_all_windows(opts.args)
+        internals.register_edit_command(opts.args)
```
Comment (tuurep) [inline plugin/registereditor.lua:5]:
```suggestion
        internals.registereditor_command(opts.args)
```

File: lua/internals.lua [hunk 4]
```diff
@@ -104,26 +110,108 @@ local function check_string_is_register(value)
...
+end
+
+-- main entry point for the :RegisterEdit user command
+M.register_edit_command = function(arg)
```
Comment (tuurep) [inline lua/internals.lua:203]:
```suggestion
-- main entry point for the :RegisterEditor user command
M.registereditor_command = function(arg)
```

**Review by tuurep — CHANGES_REQUESTED** [review]
`:RegisterEdit` => `:RegisterEditor` and related naming suggestions

**Comment by nullromo** [comment]
Applied the change.

**Comment by tuurep** [comment]
Thanks, what do you think about https://github.com/tuurep/registereditor/pull/15#discussion_r2159991543? I somewhat tested that it works the same.

With that, I think we could merge this.

**Comment by nullromo** [comment]
Forgot about that one. Thanks for the reminder. Pushed.

**Comment by tuurep** [comment]
Yeah, once again verified that it works, thank you

I'm going to do a piece of #18 so the README matches what was added

**Review by SALEH55155 — APPROVED** [review]
